src/hotspot/share/utilities/globalDefinitions_gcc.hpp

```diff
@@ -31,12 +31,23 @@
 // globally used constants & types, class (forward)
 // declarations and a few frequently used utility functions.
 
+#if defined(_AIX)
+#include <alloca.h>
+#endif
```

Comment:
I would much rather see this include added in the few places it was actually needed, rather than being
added here.

Reply:
Do we even need to include <alloca.h>? 

From the Linux man page for alloca:

By necessity, alloca() is a compiler built-in, also known as
__builtin_alloca().  By default, modern compilers automatically
translate all uses of alloca() into the built-in, but this is
forbidden if standards conformance is requested (-ansi, -std=c*),
in which case <alloca.h> is required, lest a symbol dependency be
emitted.

There are uses of it in shared code where there isn't an applicable include,
other than from globalDefinitions_xlc.hpp. So it appears all other supported
compilers do treat it as a built-in with the options we are providing, and
don't need the include. Maybe that's true for the new xlc compiler too?

Reply:
If I omit this #include <alloca.h>
I get compiler errors of the following kind
```
.../src/hotspot/share/runtime/javaThread.cpp:2222:24: error: use of undeclared identifier 'alloca'
    char* p1 = (char*) alloca(1);
                       ^

```
Of course I can do this include in every nagging file, but I thought it is simpler to keep it in the central header.

Reply:
Is the comment in front of https://github.com/openjdk/jdk/blob/51ed69a586105b707ae616f9eba898449bf9fba7/src/hotspot/os/aix/os_aix.cpp#L28 still correct? Seems like it isn't followed everywhere.

Reply:
Can `-Dalloca=__builtin_alloca` replace `#include <alloca.h>`?

Reply:
Yes I believe. I will remove the `#pragma alloca` everywhere, I will remove the `#include <alloca.h>` everywhere and I will add  `-Dalloca=__builtin_alloca` to the compile commands. If it works I will update the PR.

Reply:
In my humble opinion the inclusion of alloca.h was slightly cleaner, but I guess it doesn't matter. Out of curiosity, why do you guys prefer not including it?

Reply:
When only looking at AIX code, I think the inclusion of alloca.h was cleaner. Agreed. The new code makes AIX behave like other platforms and avoids the AIX specific part in shared code.
I could live with either version.

Reply:
That build failure in shared code does not happen with Xcode clang, gcc, or
Visual Studio, even though none of them appear to have a relevant define or
include. So the clang variant being used for AIX is different from the Xcode
clang variant (and maybe others) in its treatment of alloca. Weird!

I can also live with either the macro or the includes where needed.  I dislike
conditionally adding the include in globalDefinitions_gcc.hpp.

Reply:
Should also remove the `#pragma alloca` in os_aix.cpp.

Reply:
It was too bad that I did not see and review this change in the makefiles. :-(

While you guys could have gone either way, I strongly dislike the choice to include a redefinition in the makefiles. If this really should be done, we should introduced a new variable to carry such changes, instead of piggybacking it with the OS defines. :-( But, I don't think it should be done at all.

There are several reasons why this is a inferior solution:

1) It does not follow prior examples. We have tried hard before not do things like this, but rather pass flags as defines (e.g. `-DREDEFINE_ALLOCA` had been better)
2) It does not scale. If we start in effect allowing code in the command line, there is no clear limit anymore what should be placed in the source code files and what should be placed on the command line.
3) It messes up command lines. Keeping command lines as short as reasonable possible is a goal we try to strive for. In this case, there is also the `'` inside them (which I don't understand why), which is just begging for quoting/escaping problems, making command lines hard to copy/paste, send to different systems (like logging) etc.

I'd really like to see a follow-up PR that moves this away from the command line define and into a source code file instead.

Reply:
Can we unconditionally `#include <alloca.h>` in all files which use `alloca`? Or does that disturb any platform?

Reply:
I don't think it does. From the documentation I've checked, `#include <alloca.h>` is valid everywhere. I'm not sure why the fact that it is a compiler-builtin is relevant here. The man page for alloca on Linux says:

```
SYNOPSIS
       #include <alloca.h>

       void *alloca(size_t size);
```

which I take it as this is the formally correct way to use alloca. If it happens to work without the include file, that's just coincidence, and not a sign that we should remove `#include <alloca.h>` everywhere. @kimbarrett What do you say?

Reply:
We can not use `#include <alloca.h>` in all files which use `alloca`, because windows does not know this header. Maybe we can use `#include <alloca.h>` unconditionally in globalDefinitions_gcc.hpp, if windows will never use this file. @kimbarrett What do you say?

Reply:
That was kind of where the discussion started, and which Kim did not like. If I read him correctly, his suggestion was instead to place:
```
#if defined(_AIX)
#include <alloca.h>
#endif
```
in the files where `alloca` is needed on AIX.

Reply:
(If some of these files happen to be files which are not compiled on Windows, I assume it will not hurt to drop the ifdef guard, but then again, it can certainly be kept as well for consistency.)

Reply:
Windows does use this file, in the unofficial Windows/gcc Port. That said, I am fairly sure the Windows distribution of gcc does recognise alloca.h. It would be a little strange to unconditionally include this if only AIX needs it, though

Reply:
@magicus @TheShermanTanker @TheRealMDoerr @kimbarrett 
Let me summarize the choices we have and ask for your vote.
Julian dislikes the `-Dalloca'(size)'=__builtin_alloca'(size)'` in `flags-cflags.m4` I introduced to get rid of 
```
#if defined(_AIX)
#include <alloca.h>
#endif
```
in `globalDefinitions_gcc.hpp`. 

We have three possible solutions

1. Reintroduce
```
#if defined(_AIX)
#include <alloca.h>
#endif
```
in `globalDefinitions_gcc.hpp`. 

2. Unconditionally introduce only `#include <alloca.h>` in `globalDefinitions_gcc.hpp`. This should work for all platforms using this header including the unofficial Windows/gcc Port, although only AIX needs it. 

3. Add 
```
#if defined(_AIX)
#include <alloca.h>
#endif
```
to the sources using alloca(). These are
/hotspot/share/runtime/os.cpp
/hotspot/share/runtime/javaThread.cpp
/hotspot/share/utilities/vmError.cpp
Here we need the AIX condition, because otherwise the classic Windows Build (NTAMD64) fails.

I will implement the solution with the most likes and having no dislike.

Reply:
I don't mind all 3, though I certainly prefer 1 and 3 over 2 (The way I see it, the AIX macro check is more of a message to the programmer than it is important to the compiler, so I prefer the options that have it. However, I also don't mind if we were to go the way of option 2, this is more of a preference thing). The fact that only 3 files need it is also surprising to me, and makes option 3 seem like a good fit (Again, personal preference)

Magnus and Kim, what do you guys think?

Reply:
If there are just 3 files using alloca, I strongly prefer solution 3. I think solution 1 has already been rejected by Kim.

(Also, for the record, it was me, not Julian, who expressed dislike about the `-Dalloca'(size)'=__builtin_alloca'(size)'` change)

Reply:
I'm aware of this discussion and looking into the issues, but a personal matter has intervened and it will take
me a while to respond properly.  Maybe next week.

Reply:
I opened https://bugs.openjdk.org/browse/JDK-8330539 so we don't lose track of this, but we can keep the discussion/voting here.

Reply:
For the impatient, I suggest adopting mechanism 2, i.e. unconditionally
include <alloc.h> in globalDefinitions_gcc.hpp.

We can't include <alloca.h> in shared code, and there is a use in shared code
(in the relatively recently added JavaThread::pretouch_stack).

When I questioned whether we needed to include <alloca.h> at all, I referred
to a Linux man page I'd found on the internet (the same page mdoerr linked
to), which says (in part)

"By default, modern compilers automatically translate all uses of alloca()
into the built-in ..."

Apparently I should have kept digging, because it seems that page is
old/incorrect. A seemingly more recent Linux man page describes a different
way of handling it that is closer to what we're seeing, but still not quite
correct.

glibc's <stdlib.h> includes <alloca.h> if __USE_MISC is defined.
One of the ways __USE_MISC can become defined is if _GNU_SOURCE is defined,
and we define that for both gcc and clang toolchains.

We include <stdlib.h> in globalDefinitions_gcc.hpp. So when building with gcc,
globalDefinitions.hpp implicitly includes <alloca.h>.

The glibc definition of alloca is

#ifdef	__GNUC__
# define alloca(size)	__builtin_alloca (size)
#endif /* GCC.  */

So that explains why we don't need any explicit include of <alloca.h> when
building with gcc.  I expect there's something similar going on with Visual
Studio and Xcode/clang.  But apparently not with Open XLC clang.

Reply:
I don't think leaving out `-std=c++14` for AIX is a good solution.

Reply:
I got it. And what about simply disabling the `__STRICT_ANSI__` with
`CFLAGS_OS_DEF_JVM="-DAIX -D_LARGE_FILES -U__STRICT_ANSI__"` in flags-cflags.m4 for AIX. This worked too. The build is fine.

Reply:
So what you are saing is basically replacing
```
 CFLAGS_OS_DEF_JVM="-DAIX -Dalloca'(size)'=__builtin_alloca'(size)' -D_LARGE_FILES"
 ```
with
```
 CFLAGS_OS_DEF_JVM="-DAIX -D_LARGE_FILES -U__STRICT_ANSI__"
 ```
?

Yeah, that'll work, I guess. "strict ansi" sounds like a problematic thing to have enabled, and that it is added by `-std=c++14` sounds close to a bug in my ears. So a "workaround" where this is disabled seem reasonable.

Reply:
Yes this would be the replacement. This is our 4th way to fix the issue.
Anyone else who would prefer this too?

Reply:
Ok for me. Let's hear what @kimbarrett thinks.

Reply:
I will do after labor day and create a PR with this suggested solution in your JDK-8330539.

Reply:
I think I still prefer just unconditionally including <alloca.h> in globalDefinitions_gcc.hpp.  For gcc/clang we are using `-std=c++14` + `-D_GNU_SOURCE` instead of `-std=gnu++14`.  I forget exactly why.  I don't really want
to be messing with `__STRICT_ANSI__`.

Reply:
https://github.com/openjdk/jdk/pull/19053